api/src/apps/hooks.rs

```diff
@@ -0,0 +1,567 @@
+/*-
+ * ========================LICENSE_START=================================
+ * PREvant REST API
+ * %%
+ * Copyright (C) 2018 - 2021 aixigo AG
+ * %%
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * =========================LICENSE_END==================================
+ */
+use crate::apps::{Apps, AppsServiceError};
+use crate::models::{
+    AppName, ContainerType, Environment, EnvironmentVariable, Image, ServiceConfig,
+};
+use boa::exec::Executable;
+use boa::property::Attribute;
+use boa::syntax::ast::node::Node;
+use boa::{Context, Value};
+use secstr::SecUtf8;
+use std::collections::BTreeMap;
+use std::iter::IntoIterator;
+use std::path::{Path, PathBuf};
+
+impl Apps {
+    pub(super) async fn apply_deployment_hook(
+        &self,
+        app_name: &AppName,
+        configs: Vec<ServiceConfig>,
+    ) -> Result<Vec<ServiceConfig>, AppsServiceError> {
+        match self.config.hook("deployment") {
+            None => Ok(configs),
+            Some(hook_path) => self.parse_and_run_hook(app_name, configs, hook_path).await,
+        }
+    }
+
+    async fn parse_and_run_hook(
+        &self,
+        app_name: &AppName,
+        configs: Vec<ServiceConfig>,
+        hook_path: &Path,
+    ) -> Result<Vec<ServiceConfig>, AppsServiceError> {
+        match Self::parse_hook(hook_path).await {
+            Some(mut context) => {
+                Self::register_configs_as_global_property(&mut context, &configs);
+                context.register_global_property(
+                    "appName",
+                    Value::String(app_name.to_string().into()),
+                    Attribute::READONLY,
+                );
+
+                let transformed_configs = context
+                    .eval("deploymentHook(appName, serviceConfigs)")
```

Comment:
I'm confused why this is calles serviceConfigs, when before all I find is configs. How does this work?

Reply:
got it, its below registered as global I guess. Quite confusing code here 😅

Reply:
I agree and that was the tricky part. I created [a PR for Boa](https://github.com/boa-dev/boa/pull/1357/files) that explains how to do it. Maybe, this Boa PR clarifies it for you as well.